File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/blood_glucose_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show
        BloodGlucoseRecord,
        BloodGlucose,
        BloodGlucoseRelationToMeal,
        BloodGlucoseSpecimenSource,
        BloodGlucoseMealType,
        HealthRecordId,
        sinceV1_4_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show
        BloodGlucoseRecordDto,
        BloodGlucoseDto,
        BloodGlucoseRelationToMealDto,
        BloodGlucoseSpecimenSourceDto,
        MealTypeDto;
import 'package:meta/meta.dart' show internal;

/// Converts [BloodGlucoseRecord] to [BloodGlucoseRecordDto].
@sinceV1_4_0
@internal
extension BloodGlucoseRecordToDto on BloodGlucoseRecord {
  BloodGlucoseRecordDto toDto() {
    return BloodGlucoseRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      bloodGlucose: bloodGlucose.toDto() as BloodGlucoseDto,
      relationToMeal: relationToMeal.toDto(),
      specimenSource: specimenSource.toDto(),
      mealType: mealType.toDto(),
    );
  }
}

/// Converts [BloodGlucoseRecordDto] to [BloodGlucoseRecord].
@sinceV1_4_0
@internal
extension BloodGlucoseRecordDtoToDomain on BloodGlucoseRecordDto {
  BloodGlucoseRecord toDomain() {
    return BloodGlucoseRecord(
      id: id?.toHealthRecordId() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      bloodGlucose: bloodGlucose.toDomain() as BloodGlucose,
      relationToMeal:
          relationToMeal?.toDomain() ?? BloodGlucoseRelationToMeal.unknown,
      specimenSource:
          specimenSource?.toDomain() ?? BloodGlucoseSpecimenSource.unknown,
      mealType:
          mealType?.toBloodGlucoseMealType() ?? BloodGlucoseMealType.unknown,
    );
  }
}

/// Converts [BloodGlucoseRelationToMeal] to [BloodGlucoseRelationToMealDto].
@sinceV1_4_0
@internal
extension BloodGlucoseRelationToMealToDto on BloodGlucoseRelationToMeal {
  /// Converts [BloodGlucoseRelationToMeal] to [BloodGlucoseRelationToMealDto].
  BloodGlucoseRelationToMealDto toDto() {
    switch (this) {
      case BloodGlucoseRelationToMeal.unknown:
        return BloodGlucoseRelationToMealDto.unknown;
      case BloodGlucoseRelationToMeal.general:
        return BloodGlucoseRelationToMealDto.general;
      case BloodGlucoseRelationToMeal.fasting:
        return BloodGlucoseRelationToMealDto.fasting;
      case BloodGlucoseRelationToMeal.beforeMeal:
        return BloodGlucoseRelationToMealDto.beforeMeal;
      case BloodGlucoseRelationToMeal.afterMeal:
        return BloodGlucoseRelationToMealDto.afterMeal;
    }
  }
}

/// Converts [BloodGlucoseRelationToMealDto] to [BloodGlucoseRelationToMeal].
@sinceV1_4_0
@internal
extension BloodGlucoseRelationToMealDtoToDomain
    on BloodGlucoseRelationToMealDto {
  BloodGlucoseRelationToMeal toDomain() {
    switch (this) {
      case BloodGlucoseRelationToMealDto.unknown:
        return BloodGlucoseRelationToMeal.unknown;
      case BloodGlucoseRelationToMealDto.general:
        return BloodGlucoseRelationToMeal.general;
      case BloodGlucoseRelationToMealDto.fasting:
        return BloodGlucoseRelationToMeal.fasting;
      case BloodGlucoseRelationToMealDto.beforeMeal:
        return BloodGlucoseRelationToMeal.beforeMeal;
      case BloodGlucoseRelationToMealDto.afterMeal:
        return BloodGlucoseRelationToMeal.afterMeal;
    }
  }
}

/// Converts [BloodGlucoseSpecimenSource] to [BloodGlucoseSpecimenSourceDto].
@sinceV1_4_0
@internal
extension BloodGlucoseSpecimenSourceToDto on BloodGlucoseSpecimenSource {
  /// Converts [BloodGlucoseSpecimenSource] to [BloodGlucoseSpecimenSourceDto].
  BloodGlucoseSpecimenSourceDto toDto() {
    switch (this) {
      case BloodGlucoseSpecimenSource.unknown:
        return BloodGlucoseSpecimenSourceDto.unknown;
      case BloodGlucoseSpecimenSource.interstitialFluid:
        return BloodGlucoseSpecimenSourceDto.interstitialFluid;
      case BloodGlucoseSpecimenSource.capillaryBlood:
        return BloodGlucoseSpecimenSourceDto.capillaryBlood;
      case BloodGlucoseSpecimenSource.plasma:
        return BloodGlucoseSpecimenSourceDto.plasma;
      case BloodGlucoseSpecimenSource.serum:
        return BloodGlucoseSpecimenSourceDto.serum;
      case BloodGlucoseSpecimenSource.tears:
        return BloodGlucoseSpecimenSourceDto.tears;
      case BloodGlucoseSpecimenSource.wholeBlood:
        return BloodGlucoseSpecimenSourceDto.wholeBlood;
    }
  }
}

/// Converts [BloodGlucoseSpecimenSourceDto] to [BloodGlucoseSpecimenSource].
@sinceV1_4_0
@internal
extension BloodGlucoseSpecimenSourceDtoToDomain
    on BloodGlucoseSpecimenSourceDto {
  BloodGlucoseSpecimenSource toDomain() {
    switch (this) {
      case BloodGlucoseSpecimenSourceDto.unknown:
        return BloodGlucoseSpecimenSource.unknown;
      case BloodGlucoseSpecimenSourceDto.interstitialFluid:
        return BloodGlucoseSpecimenSource.interstitialFluid;
      case BloodGlucoseSpecimenSourceDto.capillaryBlood:
        return BloodGlucoseSpecimenSource.capillaryBlood;
      case BloodGlucoseSpecimenSourceDto.plasma:
        return BloodGlucoseSpecimenSource.plasma;
      case BloodGlucoseSpecimenSourceDto.serum:
        return BloodGlucoseSpecimenSource.serum;
      case BloodGlucoseSpecimenSourceDto.tears:
        return BloodGlucoseSpecimenSource.tears;
      case BloodGlucoseSpecimenSourceDto.wholeBlood:
        return BloodGlucoseSpecimenSource.wholeBlood;
    }
  }
}

/// Converts [BloodGlucoseMealType] to [MealTypeDto].
@sinceV1_4_0
@internal
extension BloodGlucoseMealTypeToDto on BloodGlucoseMealType {
  MealTypeDto toDto() {
    switch (this) {
      case BloodGlucoseMealType.unknown:
        return MealTypeDto.unknown;
      case BloodGlucoseMealType.breakfast:
        return MealTypeDto.breakfast;
      case BloodGlucoseMealType.lunch:
        return MealTypeDto.lunch;
      case BloodGlucoseMealType.dinner:
        return MealTypeDto.dinner;
      case BloodGlucoseMealType.snack:
        return MealTypeDto.snack;
    }
  }
}

/// Converts [MealTypeDto] to [BloodGlucoseMealType].
@sinceV1_4_0
@internal
extension MealTypeDtoToBloodGlucoseMealType on MealTypeDto {
  BloodGlucoseMealType toBloodGlucoseMealType() {
    switch (this) {
      case MealTypeDto.unknown:
        return BloodGlucoseMealType.unknown;
      case MealTypeDto.breakfast:
        return BloodGlucoseMealType.breakfast;
      case MealTypeDto.lunch:
        return BloodGlucoseMealType.lunch;
      case MealTypeDto.dinner:
        return BloodGlucoseMealType.dinner;
      case MealTypeDto.snack:
        return BloodGlucoseMealType.snack;
    }
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/BloodGlucoseRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.MealType
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseRelationToMealDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseSpecimenSourceDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts [BloodGlucoseRelationToMealDto] to Health Connect int constant.
 */
internal fun BloodGlucoseRelationToMealDto.toHealthConnect(): Int = when (this) {
    BloodGlucoseRelationToMealDto.UNKNOWN -> BloodGlucoseRecord.RELATION_TO_MEAL_UNKNOWN
    BloodGlucoseRelationToMealDto.GENERAL -> BloodGlucoseRecord.RELATION_TO_MEAL_GENERAL
    BloodGlucoseRelationToMealDto.FASTING -> BloodGlucoseRecord.RELATION_TO_MEAL_FASTING
    BloodGlucoseRelationToMealDto.BEFORE_MEAL -> BloodGlucoseRecord.RELATION_TO_MEAL_BEFORE_MEAL
    BloodGlucoseRelationToMealDto.AFTER_MEAL -> BloodGlucoseRecord.RELATION_TO_MEAL_AFTER_MEAL
}

/**
 * Converts Health Connect int constant to [BloodGlucoseRelationToMealDto].
 */
internal fun Int.toBloodGlucoseRelationToMealDto(): BloodGlucoseRelationToMealDto = when (this) {
    BloodGlucoseRecord.RELATION_TO_MEAL_GENERAL -> BloodGlucoseRelationToMealDto.GENERAL
    BloodGlucoseRecord.RELATION_TO_MEAL_FASTING -> BloodGlucoseRelationToMealDto.FASTING
    BloodGlucoseRecord.RELATION_TO_MEAL_BEFORE_MEAL -> BloodGlucoseRelationToMealDto.BEFORE_MEAL
    BloodGlucoseRecord.RELATION_TO_MEAL_AFTER_MEAL -> BloodGlucoseRelationToMealDto.AFTER_MEAL
    else -> BloodGlucoseRelationToMealDto.UNKNOWN
}

/**
 * Converts [BloodGlucoseSpecimenSourceDto] to Health Connect int constant.
 */
internal fun BloodGlucoseSpecimenSourceDto.toHealthConnect(): Int = when (this) {
    BloodGlucoseSpecimenSourceDto.UNKNOWN -> BloodGlucoseRecord.SPECIMEN_SOURCE_UNKNOWN
    BloodGlucoseSpecimenSourceDto.INTERSTITIAL_FLUID ->
        BloodGlucoseRecord.SPECIMEN_SOURCE_INTERSTITIAL_FLUID

    BloodGlucoseSpecimenSourceDto.CAPILLARY_BLOOD ->
        BloodGlucoseRecord.SPECIMEN_SOURCE_CAPILLARY_BLOOD

    BloodGlucoseSpecimenSourceDto.PLASMA -> BloodGlucoseRecord.SPECIMEN_SOURCE_PLASMA
    BloodGlucoseSpecimenSourceDto.SERUM -> BloodGlucoseRecord.SPECIMEN_SOURCE_SERUM
    BloodGlucoseSpecimenSourceDto.TEARS -> BloodGlucoseRecord.SPECIMEN_SOURCE_TEARS
    BloodGlucoseSpecimenSourceDto.WHOLE_BLOOD -> BloodGlucoseRecord.SPECIMEN_SOURCE_WHOLE_BLOOD
}

/**
 * Converts Health Connect int constant to [BloodGlucoseSpecimenSourceDto].
 */
internal fun Int.toBloodGlucoseSpecimenSourceDto(): BloodGlucoseSpecimenSourceDto = when (this) {
    BloodGlucoseRecord.SPECIMEN_SOURCE_INTERSTITIAL_FLUID ->
        BloodGlucoseSpecimenSourceDto.INTERSTITIAL_FLUID

    BloodGlucoseRecord.SPECIMEN_SOURCE_CAPILLARY_BLOOD ->
        BloodGlucoseSpecimenSourceDto.CAPILLARY_BLOOD

    BloodGlucoseRecord.SPECIMEN_SOURCE_PLASMA -> BloodGlucoseSpecimenSourceDto.PLASMA
    BloodGlucoseRecord.SPECIMEN_SOURCE_SERUM -> BloodGlucoseSpecimenSourceDto.SERUM
    BloodGlucoseRecord.SPECIMEN_SOURCE_TEARS -> BloodGlucoseSpecimenSourceDto.TEARS
    BloodGlucoseRecord.SPECIMEN_SOURCE_WHOLE_BLOOD -> BloodGlucoseSpecimenSourceDto.WHOLE_BLOOD
    else -> BloodGlucoseSpecimenSourceDto.UNKNOWN
}

/**
 * Converts a Health Connect [BloodGlucoseRecord] to a [BloodGlucoseRecordDto].
 */
internal fun BloodGlucoseRecord.toDto(): BloodGlucoseRecordDto = BloodGlucoseRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    bloodGlucose = level.toDto(),
    relationToMeal = relationToMeal.toBloodGlucoseRelationToMealDto(),
    specimenSource = specimenSource.toBloodGlucoseSpecimenSourceDto(),
    mealType = mealType.toMealTypeDto(),
)

/**
 * Converts a [BloodGlucoseRecordDto] to a Health Connect [BloodGlucoseRecord].
 */
internal fun BloodGlucoseRecordDto.toHealthConnect(): BloodGlucoseRecord = BloodGlucoseRecord(
    time = Instant.ofEpochMilli(time),
    zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    metadata = metadata.toHealthConnect(),
    level = bloodGlucose.toHealthConnect(),
    relationToMeal = relationToMeal?.toHealthConnect()
        ?: BloodGlucoseRecord.RELATION_TO_MEAL_UNKNOWN,
    specimenSource = specimenSource?.toHealthConnect()
        ?: BloodGlucoseRecord.SPECIMEN_SOURCE_UNKNOWN,
    mealType = mealType?.toHealthConnect() ?: MealType.MEAL_TYPE_UNKNOWN,
)
